### ISSUE TITLE
can rusnc support yama: does not work with ptrace_scope == 2?

### ISSUE THREAD

**Issue by sitano** [google/gvisor#8989]
### Description

I have run into an issue with Amazon Linux where Yama is configured in mode == 2 (https://www.kernel.org/doc/Documentation/security/Yama.txt):

```
# cat /proc/sys/kernel/yama/ptrace_scope 
2

# runsc create ivan
creating container: cannot create sandbox: cannot read client sync file: waiting for sandbox to start: EOF
```

in debug log

```
panic: unable to attach: operation not permitted
```

how to fix:

```
# echo 1 > /proc/sys/kernel/yama/ptrace_scope
# runsc create ivan
done
```

mode 2 requires children to explicitly ask for the ptrace:

> 2 - admin-only attach: only processes with CAP_SYS_PTRACE may use ptrace with PTRACE_ATTACH, or through children calling PTRACE_TRACEME.

I am curious if we can make it potentially compatible by offering PTRACE_TRACEME?

### Do you have a specific solution in mind?

```
# echo 1 > /proc/sys/kernel/yama/ptrace_scope
# runsc create ivan
done
```

### Environment

Amazon EKS, privileged pods, completely disabled SElinux, root, 5.15.110-70.141.amzn2.x86_64 (Debian GNU/Linux 11 (bullseye)).

**Comment by nixprime** [comment]
> 2 - admin-only attach: only processes with CAP_SYS_PTRACE may use ptrace with PTRACE_ATTACH, or through children calling PTRACE_TRACEME.

IIUC, this means that `CAP_SYS_PTRACE` is required for both `PTRACE_ATTACH` and `PTRACE_TRACEME`, not that `PTRACE_TRACEME` is exempt from the check; see [the check for the latter](https://github.com/torvalds/linux/blob/e2065b8c1b0120d47b327364a1a09090bdc11f31/security/yama/yama_lsm.c#L406).

[runsc will retain `CAP_SYS_PTRACE` if it starts with `CAP_SYS_PTRACE` and a ptrace-using platform is in use](https://github.com/google/gvisor/blob/7e32a238387ba3945d15855b013231759d21d8e4/runsc/cmd/boot.go#L301), but in your example you are running runsc as a regular non-root program, so it gets no capabilities. I think it should work if runsc is provided `CAP_SYS_PTRACE`, e.g. by running it as root or by using [setcap](https://man7.org/linux/man-pages/man8/setcap.8.html).

**Comment by sitano** [comment]
@nixprime I am sorry if you were confused by `$`, but

> but in your example you are running runsc as a regular non-root program, so it gets no capabilities. I think it should work if runsc is provided CAP_SYS_PTRACE, e.g. by running it as root or by using [setcap](https://man7.org/linux/man-pages/man8/setcap.8.html).

I am running it under `root` with a full set of unrestricted privileges. The point of this Issue is that `gvisor/runsc` does not work under YAMA mode 2. However, I misread what mode 2 is about - I thought it always requires PTRACE_TRACEME. The reference tho states that the CAP_SYS_PTRACE must be enough. However, it does not work rn. It looked like the CAP_SYS_PTRACE is not enough, but now I am not sure what is the cause.

To be specific, I am using Amazon EKS environment with fully privileged pods with Amazon Linux kernels `5.15.110-70.141.amzn2.x86_64` (Debian GNU/Linux 11 (bullseye)). It is pods with root (`CapEff: 000001ffffffffff`), all privs are there and all the securities are disabled.

**Comment by sitano** [comment]
@nixprime ok, I think the reason is that line https://github.com/torvalds/linux/blob/e2065b8c1b0120d47b327364a1a09090bdc11f31/security/yama/yama_lsm.c#L377

> !ns_capable(__task_cred(child)->user_ns, CAP_SYS_PTRACE)) rc = -EPERM;

If I get it correctly, it say: the target (child) process must have a CAP_SYS_PTRACE in its User Namespace. As we can see it does not have it.

Kernel stacktrace proof:

```
# perf ftrace  -a -v -G __x64_sys_ptrace
# yama == 2
# runsc create test

function_graph tracer is used
# tracer: function_graph
#
# CPU  DURATION                  FUNCTION CALLS
# |     |   |                     |   |   |   |
 4)               |  __x64_sys_ptrace() {
 4)               |    ptrace_attach() {
 4)   0.171 us    |      mutex_lock_interruptible();
 4)   0.110 us    |      _raw_spin_lock();
 4)               |      __ptrace_may_access() {
 4)   0.091 us    |        rcu_read_unlock_strict();
 4)               |        security_ptrace_access_check() {
 4)               |          cap_ptrace_access_check() {
 4)   0.090 us    |            rcu_read_unlock_strict();
 4)   0.290 us    |          }
 4)               |          yama_ptrace_access_check() { <-------------
 4)               |            yama_ptrace_access_check.part.0() {
 4)               |              ns_capable() {
 4)               |                security_capable() {
 4)   0.101 us    |                  cap_capable();
 4)   0.409 us    |                }
 4)   0.590 us    |              }
 4)   0.100 us    |              rcu_read_unlock_strict();
 4)               |              report_access() {
 4)               |                kmem_cache_alloc_trace() {
 4)   0.090 us    |                  should_failslab();
 4)   0.400 us    |                }
 4)               |                task_work_add() {
 4)   0.100 us    |                  kick_process();
 4)   0.320 us    |                }
 4)   1.010 us    |              }
 4)   2.089 us    |            }
 4)   2.371 us    |          }
 4)   3.341 us    |        }
 4)   3.821 us    |      }
 4)   0.090 us    |      mutex_unlock();
 4)   4.729 us    |    }
 4) + 20.840 us   |  }
```

**Comment by sitano** [comment]
From what I can understand, it will never work for runsc/PTRACE_ATTACH/YAMA_SCOPE_CAPABILITY as per:

```
		case YAMA_SCOPE_CAPABILITY:
			rcu_read_lock();
			if (!ns_capable(__task_cred(child)->user_ns, CAP_SYS_PTRACE))
				rc = -EPERM;
			rcu_read_unlock();
			break;
```

as the security model would never allow having CAP_SYS_PTRACE in the child user namespace generally - it does not make sense - we want to restrict target sandboxes, not give away permissions. But with the TRACEME the check looks better:

```
	case YAMA_SCOPE_CAPABILITY:
		if (!has_ns_capability(parent, current_user_ns(), CAP_SYS_PTRACE))
			rc = -EPERM;
		break;
``` 

as it looks like its enough for the parent to just be an owner of the descendant user namespace as per https://github.com/torvalds/linux/blob/44c026a73be8038f03dbdeef028b642880cf1511/security/commoncap.c#L91-L92.

**Comment by sitano** [comment]
Pardon, here I must be wrong https://github.com/google/gvisor/issues/8989#issuecomment-1556975111. Those 2 checks must be equivalent afaiu - attaching process has proper cap in target process user ns or is the owner of it and in effective user. Some piece of this puzzle is missing.

**Comment by sitano** [comment]
@nixprime in addition to that, I am finding runsc strace interesting, as it seems to be failing at the same point, even though due to the different reason - because strace uses ptrace, so the traced process can't be ptrace-attached. But that is giving some interesting context on the structure of the processes that gvisor/runsc is using:

[strace_gvisor_yama2.log](https://github.com/google/gvisor/files/11545766/strace_gvisor_yama2.log)

As you can see 579th process (runsc-sandbox) clones into 605th which in its turn clones into 612th and then 605th calls PTRACE_ATTACH on 612th. 

    579 (in new user ns) -clone-> 605 -clone->  612,
                                  605 -attach-> 612

But here we have 2 interesting things AFAIGI:

- all those processes 579, 605, 612 are in the SAME user namespace so parenting condition would not work from cap_capability
- AFAIGI 579 does not have PTRACE capability after executing EXECVE syscall due to inheritance cap = SYS_ADMIN
- e{u,g}id is not anywhere near to the child->user_ns->owner.

So I expect this line is executed and fails https://github.com/torvalds/linux/blob/44c026a73be8038f03dbdeef028b642880cf1511/security/commoncap.c#L77-L78:

		/* Do we have the necessary capabilities? */
		if (ns == cred->user_ns)
			return cap_raised(cred->cap_effective, cap) ? 0 : -EPERM;

Because:

    /////  inheritable=1<<CAP_SYS_CHROOT|1<<CAP_SYS_ADMIN
    [pid 1547579] capset({version=_LINUX_CAPABILITY_VERSION_3, pid=0}, {effective=1<<CAP_CHOWN|1<<CAP_DAC_OVERRIDE|1<<CAP_DAC_READ_SEARCH|1<<CAP_FOWNER|1<<CAP_FSETID|1<<CAP_KILL|1<<CAP_SETGID|1<<CAP_SETUID|1<<CAP_SETPCAP|1<<CAP_LINUX_IMMUTABLE|1<<CAP_NET_BIND_SERVICE|1<<CAP_NET_BROADCAST|1<<CAP_NET_ADMIN|1<<CAP_NET_RAW|1<<CAP_IPC_LOCK|1<<CAP_IPC_OWNER|1<<CAP_SYS_MODULE|1<<CAP_SYS_RAWIO|1<<CAP_SYS_CHROOT|1<<CAP_SYS_PTRACE|1<<CAP_SYS_PACCT|1<<CAP_SYS_ADMIN|1<<CAP_SYS_BOOT|1<<CAP_SYS_NICE|1<<CAP_SYS_RESOURCE|1<<CAP_SYS_TIME|1<<CAP_SYS_TTY_CONFIG|1<<CAP_MKNOD|1<<CAP_LEASE|1<<CAP_AUDIT_WRITE|1<<CAP_AUDIT_CONTROL|1<<CAP_SETFCAP|1<<CAP_MAC_OVERRIDE|1<<CAP_MAC_ADMIN|1<<CAP_SYSLOG|1<<CAP_WAKE_ALARM|1<<CAP_BLOCK_SUSPEND|1<<CAP_AUDIT_READ|1<<CAP_PERFMON|1<<CAP_BPF|1<<CAP_CHECKPOINT_RESTORE, permitted=1<<CAP_CHOWN|1<<CAP_DAC_OVERRIDE|1<<CAP_DAC_READ_SEARCH|1<<CAP_FOWNER|1<<CAP_FSETID|1<<CAP_KILL|1<<CAP_SETGID|1<<CAP_SETUID|1<<CAP_SETPCAP|1<<CAP_LINUX_IMMUTABLE|1<<CAP_NET_BIND_SERVICE|1<<CAP_NET_BROADCAST|1<<CAP_NET_ADMIN|1<<CAP_NET_RAW|1<<CAP_IPC_LOCK|1<<CAP_IPC_OWNER|1<<CAP_SYS_MODULE|1<<CAP_SYS_RAWIO|1<<CAP_SYS_CHROOT|1<<CAP_SYS_PTRACE|1<<CAP_SYS_PACCT|1<<CAP_SYS_ADMIN|1<<CAP_SYS_BOOT|1<<CAP_SYS_NICE|1<<CAP_SYS_RESOURCE|1<<CAP_SYS_TIME|1<<CAP_SYS_TTY_CONFIG|1<<CAP_MKNOD|1<<CAP_LEASE|1<<CAP_AUDIT_WRITE|1<<CAP_AUDIT_CONTROL|1<<CAP_SETFCAP|1<<CAP_MAC_OVERRIDE|1<<CAP_MAC_ADMIN|1<<CAP_SYSLOG|1<<CAP_WAKE_ALARM|1<<CAP_BLOCK_SUSPEND|1<<CAP_AUDIT_READ|1<<CAP_PERFMON|1<<CAP_BPF|1<<CAP_CHECKPOINT_RESTORE, inheritable=1<<CAP_SYS_CHROOT|1<<CAP_SYS_ADMIN} <unfinished ...>
    [pid 1547579] execve("/proc/self/exe", ["runsc-sandbox", "--root=/var/run/runsc", "boot", "--bundle=/tmp/123/vm", "--setup-root", "--cpu-num", "8", "--total-memory", "268435456", "--io-fds=3", "--mounts-fd=4", "--start-sync-fd=5", "--controller-fd=6", "--spec-fd=7", "--stdio-fds=8", "--stdio-fds=9", "--stdio-fds=10", "ivan"], ["LS_COLORS=rs=0:di=01;34:ln=01;36"..., "PATH=/home/sitano/.wasmer/bin:/h"..., "DISPLAY=:0", "LANG=en_US.UTF-8", "XAUTHORITY=/home/sitano/.Xauthor"..., "TERM=xterm-256color", "COLORTERM=truecolor", "LC_CTYPE=en_US.UTF-8", "MAIL=/var/mail/root", "LOGNAME=root", "USER=root", "HOME=/root", "SHELL=/bin/bash", "SUDO_COMMAND=/usr/sbin/strace -v"..., "SUDO_USER=sitano", "SUDO_UID=1000", "SUDO_GID=1001"] <unfinished ...>
[pid 1547605] clone(child_stack=0xc000110000, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS <unfinished ...>
 <unfinished ...>
    ...
    [pid 1547579] clone(child_stack=0xc00007a000, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS <unfinished ...>
    [pid 1547604] gettid(strace: Process 1547605 attached
    )                  = 1547604
    [pid 1547605] gettid( <unfinished ...>
    [pid 1547579] <... clone resumed>, tls=0xc000068490) = 1547605
    [pid 1547605] <... gettid resumed>)     = 1547605
    ...
    [pid 1547605] <... clone resumed>, tls=0xc0004ce490) = 1547611
    [pid 1547611] gettid( <unfinished ...>
    [pid 1547611] <... gettid resumed>)     = 1547611
    [pid 1547611] gettid()                  = 1547611
    [pid 1547605] getpid( <unfinished ...>
    [pid 1547605] <... getpid resumed>)     = 1547579
    [pid 1547605] clone(child_stack=NULL, flags=0 <unfinished ...>
     <unfinished ...>
    [pid 1547605] <... clone resumed>)      = 1547612
    [pid 1547612] setsid( <unfinished ...>
    [pid 1547612] <... setsid resumed>)     = 1547612
    [pid 1547605] wait4(1547612,  <unfinished ...>
    [pid 1547612] prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) = 0
    [pid 1547612] arch_prctl(ARCH_SET_CPUID, 0 <unfinished ...>
    [pid 1547612] <... arch_prctl resumed>) = 0
    [pid 1547612] prctl(PR_SET_PDEATHSIG, SIGKILL) = 0
    [pid 1547612] getppid()                 = 1547579
    [pid 1547612] getpid()                  = 1547612
    [pid 1547612] kill(1547612, SIGSTOP)    = 0
    [pid 1547612] --- SIGSTOP {si_signo=SIGSTOP, si_code=SI_USER, si_pid=1547612, si_uid=65534} ---
    [pid 1547612] --- stopped by SIGSTOP ---
    [pid 1547605] <... wait4 resumed>[{WIFSTOPPED(s) && WSTOPSIG(s) == SIGSTOP}], WSTOPPED|__WALL, NULL) = 1547612
    [pid 1547605] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_STOPPED, si_pid=1547612, si_uid=65534, si_status=SIGSTOP, si_utime=0, si_stime=0} ---
    [pid 1547605] rt_sigreturn({mask=[]})   = 1547612
    [pid 1547605] ptrace(PTRACE_ATTACH, 1547612 <unfinished ...>
    [pid 1547604] <... restart_syscall resumed>) = 0
    [pid 1547605] <... ptrace resumed>)     = -1 EPERM (Operation not permitted)

**Comment by sitano** [comment]
ok, I used a bit of an old release for my experiments.

however, at least for `runsc version release-20230327.0-dirty` I confirm that the runsc-sandbox process does not have a CAP_SYS_PTRACE permission in its Effective caps:

```
root     1786691  0.0  0.0  19740  8192 pts/12   S+   12:23   0:00  |       \_ sudo ./runsc --debug --debug-log=1.log create ivan
root     1786692  0.0  0.0 738116 21792 pts/12   Sl+  12:23   0:00  |           \_ ./runsc --debug --debug-log=1.log create ivan
root     1786698  0.0  0.0 738116 20640 ?        Ssl  12:23   0:00  |               \_ runsc-gofer --root=/var/run/runsc --debug=true --debug-log=1.log --debug-log-fd=3 gofer --bundle /tmp/123/vm --spec-fd=4 --mounts-fd=5 --io-fds=6 --apply-caps=false --setup-root=false --sync-userns-fd=-1 --proc-mount-sync-fd=14
nobody   1786702  0.0  0.0 740044 24224 ?        Ssl  12:23   0:00  |               \_ runsc-sandbox --root=/var/run/runsc --debug=true --debug-log=1.log --debug-log-fd=3 boot --proc-mount-sync-fd=19 --product-name System Product Name --bundle=/tmp/123/vm --cpu-num 8 --total-memory 268435456 --io-fds=4 --mounts-fd=5 --start-sync-fd=6 --controller-fd=7 --spec-fd=8 --stdio-fds=9 --stdio-fds=10 --stdio-fds=11 ivan
nobody   1786708  0.0  0.0 738116 19104 ?        Sl   12:23   0:00  |                   \_ /proc/self/exe umount --sync-fd=3 /proc
nobody   1786733  0.0  0.0 739724  6532 ?        Ts   12:23   0:00  |                   \_ runsc-sandbox --root=/var/run/runsc --debug=true --debug-log=1.log --debug-log-fd=3 boot --proc-mount-sync-fd=19 --product-name System Product Name --bundle=/tmp/123/vm --cpu-num 8 --total-memory 268435456 --io-fds=4 --mounts-fd=5 --start-sync-fd=6 --controller-fd=7 --spec-fd=8 --stdio-fds=9 --stdio-fds=10 --stdio-fds=11 ivan
sitano      5940  0.0  0.0 234960  7612 ?        Sl   May16   0:23  \_ /usr/lib/at-spi2-registryd --use-gnome-session
sitano     18710  0.0  0.0 432656 16768 ?        Ssl  May16   0:10  \_ /usr/lib/gvfs-udisks2-volume-monitor

$ cat /proc/1786702/status
Name:	exe
Umask:	0022
State:	S (sleeping)
Tgid:	1786702
Ngid:	0
Pid:	1786702
PPid:	1786692
TracerPid:	0
Uid:	65534	65534	65534	65534
Gid:	65534	65534	65534	65534
FDSize:	512
Groups:	0 
NStgid:	1786702
NSpid:	1786702
NSpgid:	1786702
NSsid:	1786702
...
Threads:	9
...
CapInh:	0000000000240000
CapPrm:	0000000000240000
CapEff:	0000000000240000
CapBnd:	000001ffffffffff
CapAmb:	0000000000240000
NoNewPrivs:	0
Seccomp:	0
Seccomp_filters:	0
...

W0524 12:28:56.592559  1790323 subprocess.go:295] Sandbox has_ptrace = %!d(bool=false)
W0524 12:28:56.592686  1790323 subprocess.go:296] Sandbox pid = 1790323, ppid = 1790313, euid = 65534, target tid = 1790354
W0524 12:28:56.592715  1790323 subprocess.go:297] Starting sleeping for 10 minutes
```

So without CAP_SYS_PTRACE a runsc-sandbox with yama == 2 basically can't do an ATTACH @nixprime

**Comment by sitano** [comment]
In the newer versions than release-20230327.0 thought I see the PTRACE capability is present in logs for Sandbox:

```
I0524 12:18:40.414136  1744800 capability.go:66] Capabilities applied: { effective="chown, dac_override, dac_read_search, fowner, fsetid, sys_ptrace" permitted="chown, dac_override, dac_read_search, fowner, fsetid, sys_ptrace" inheritable="empty" bounding="chown, dac_override, dac_read_search, fowner, fsetid, sys_ptrace" }
I0524 12:18:40.414179  1744800 cmd.go:74] Execve "/proc/self/exe" again, bye!
I0524 12:18:40.469654  1744800 main.go:224] ***************************
I0524 12:18:40.469680  1744800 main.go:225] Args: [runsc-sandbox --root=/var/run/runsc --debug=true --debug-log=2.log --debug-log-fd=3 boot --proc-mount-sync-fd=19 --product-name System Product Name --bundle=/tmp/123/vm --overlay-mediums=0 --cpu-num 8 --total-memory 268435456 --io-fds=4 --mounts-fd=5 --start-sync-fd=6 --controller-fd=7 --spec-fd=8 --stdio-fds=9 --stdio-fds=10 --stdio-fds=11 ivan]
```

**Comment by sitano** [comment]
ok, this is fixed in release-20230522.0, and present in release-20230517.0. so I am closing.
